proxy.go

```diff
@@ -62,51 +60,39 @@ func (vs *version) proxy(w http.ResponseWriter, r *http.Request, peers []string)
 				outstanding -= 1
 				continue
 			} else {
-				return readResponse(res)
+				return res.resp, res.peer, nil
 			}
-		case <-totalTimeout.C:
-			return nil, errProxyTimeout
+		case <-ctx.Done():
+			return nil, "", errProxyTimeout
 		case <-stageTimeout.C:
 		}
 	}
 
-	return nil, errNoAvailablePeers
+	return nil, "", errNoAvailablePeers
 }
 
-func (vs *version) proxyAttempt(url string, res chan proxyResponse, cancel chan struct{}) {
+func (vs *version) proxyAttempt(ctx context.Context, peer string, res chan proxyResponse) {
+	url := fmt.Sprintf("http://%s/%s/%s?proxy=%s", peer, vs.db, vs.name)
```

Comment:
This seems like it's vulnerable to all sorts of fun injection attacks. Maybe construct a `url.URL` object by hand?